cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,747 @@
+        get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+        get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q) + accumulate_pair_size,
+        generate_p_r_q_r<vertex_t,
+                         edge_t,
+                         decltype(get_dataframe_buffer_begin(invalid_edges_buffer))>{
+          1,
+          raft::device_span<size_t const>(intersection_offsets.data(), intersection_offsets.size()),
+          raft::device_span<vertex_t const>(intersection_indices.data(),
+                                            intersection_indices.size()),
+          get_dataframe_buffer_begin(invalid_edges_buffer)  // FIXME: verify this is accurate
+        });
+
+      thrust::for_each(handle.get_thrust_policy(),
+                       thrust::make_counting_iterator<edge_t>(0),
+                       thrust::make_counting_iterator<edge_t>(num_edge_exists),
+                       unroll_edge<vertex_t, edge_t, decltype(edge_first)>{
+                         raft::device_span<vertex_t>(num_triangles.data(), num_triangles.size()),
+                         get_dataframe_buffer_begin(vertex_pair_buffer_q_r_edge_p_q),
+                         edge_first,
+                         edge_last,
+                       });
+
+      // Put edges with triangle count == 0 in the second partition
+      // FIXME: revisit all the 'stable_partition' and only used them
+      // when necessary otherwise simply call 'thrust::partition'
+      // Stable_parition is needed because we want to keep src and dst sorted
+      // so that we don't need to sort it again.
+      auto last_edge_with_triangles =
+        thrust::stable_partition(handle.get_thrust_policy(),
+                                 edge_triangle_count_pair_first,
+                                 edge_triangle_count_pair_first + edgelist_srcs.size(),
+                                 [] __device__(auto edge_to_num_triangles) {
+                                   return thrust::get<1>(edge_to_num_triangles) > 0;
+                                 });
+
+      auto last_edge_with_triangles_idx =
+        thrust::distance(edge_triangle_count_pair_first, last_edge_with_triangles);
+
+      // resize the 'edgelist_srcs' and 'edgelsit_dst'
+      edgelist_srcs.resize(last_edge_with_triangles_idx, handle.get_stream());
+      edgelist_dsts.resize(last_edge_with_triangles_idx, handle.get_stream());
+      num_triangles.resize(last_edge_with_triangles_idx, handle.get_stream());
+
+      auto invalid_edge_last_p_q =
+        thrust::stable_partition(handle.get_thrust_policy(),
+                                 edge_triangle_count_pair_first,
+                                 edge_triangle_count_pair_first + num_triangles.size(),
+                                 [k] __device__(auto e) {
+                                   auto num_triangles = thrust::get<1>(e);
+                                   return num_triangles < k - 2;
+                                 });
```

Comment:
We're running this stable partition at the beginning of the while loop. No need to perform this here.

Reply:
Right this is an unnecessary partitioning call.